File: src/commonMain/kotlin/com/scoreo/application/EloCalculator.kt
package com.scoreo.application

import com.scoreo.domain.model.GameType
import com.scoreo.domain.model.Match
import kotlin.math.pow
import kotlin.math.roundToInt

private const val K = 32

class EloCalculator {
    fun compute(matches: List<Match>, gameTypes: Map<String, GameType>): Map<String, Int> {
        val elo = mutableMapOf<String, Int>()
        val sorted = matches.sortedBy { it.date }

        for (match in sorted) {
            val gt = gameTypes[match.gameTypeId] ?: continue
            val winners = match.getWinners(gt).toSet()
            if (winners.isEmpty()) continue
            val participants = match.playerScores.map { it.playerId }
            val preElo = elo.toMap()
            val kNorm = K.toDouble() / (participants.size - 1)
            val deltas = mutableMapOf<String, Int>()

            for (winner in winners) {
                for (loser in participants) {
                    if (loser in winners) continue
                    val rW = preElo[winner] ?: 1200
                    val rL = preElo[loser] ?: 1200
                    val eW = 1.0 / (1.0 + 10.0.pow((rL - rW) / 400.0))
                    val eL = 1.0 / (1.0 + 10.0.pow((rW - rL) / 400.0))
                    deltas[winner] = (deltas[winner] ?: 0) + (kNorm * (1.0 - eW)).roundToInt()
                    deltas[loser] = (deltas[loser] ?: 0) + (kNorm * (0.0 - eL)).roundToInt()
                }
            }
            for ((pid, d) in deltas) {
                elo[pid] = (preElo[pid] ?: 1200) + d
            }
        }

        return elo
    }
}

File: src/application/eloCalculator.ts
import type { GameType } from '../domain/model/gameType'
import { getWinners, type Match } from '../domain/model/match'

const K = 32
const DEFAULT_RATING = 1200

export interface EloSnapshot {
  matchId: string
  date: number
  ratings: Map<string, number>
}

export class EloCalculator {
  compute(matches: Match[], gameTypes: Map<string, GameType>): Map<string, number> {
    const history = this.computeHistory(matches, gameTypes)
    return history.length > 0 ? history[history.length - 1].ratings : new Map()
  }

  computeHistory(matches: Match[], gameTypes: Map<string, GameType>): EloSnapshot[] {
    const elo = new Map<string, number>()
    const sorted = [...matches].sort((a, b) => a.date - b.date)
    const history: EloSnapshot[] = []

    for (const match of sorted) {
      const gt = gameTypes.get(match.gameTypeId)
      if (!gt) continue
      const winners = new Set(getWinners(match, gt))
      if (winners.size === 0) continue
      const participants = match.playerScores.map((s) => s.playerId)
      if (participants.length < 2) continue
      const preElo = new Map(elo)
      const kNorm = K / (participants.length - 1)
      const deltas = new Map<string, number>()

      for (const winner of winners) {
        for (const loser of participants) {
          if (winners.has(loser)) continue
          const rW = preElo.get(winner) ?? DEFAULT_RATING
          const rL = preElo.get(loser) ?? DEFAULT_RATING
          const eW = 1 / (1 + 10 ** ((rL - rW) / 400))
          const eL = 1 / (1 + 10 ** ((rW - rL) / 400))
          deltas.set(winner, (deltas.get(winner) ?? 0) + Math.round(kNorm * (1 - eW)))
          deltas.set(loser, (deltas.get(loser) ?? 0) + Math.round(kNorm * (0 - eL)))
        }
      }
      for (const [pid, d] of deltas) {
        elo.set(pid, (preElo.get(pid) ?? DEFAULT_RATING) + d)
      }

      history.push({ matchId: match.id, date: match.date, ratings: new Map(elo) })
    }

    return history
  }
}
